hibernate-core/src/test/java/org/hibernate/test/hql/HibernateFirstResultMaxResultsTest.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.test.hql;
+
+import java.util.List;
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+
+import org.hibernate.Query;
+import org.hibernate.Session;
+
+import org.hibernate.testing.TestForIssue;
+import org.hibernate.testing.junit4.BaseNonConfigCoreFunctionalTestCase;
+import org.junit.Test;
+
+import static org.hibernate.testing.transaction.TransactionUtil.doInHibernate;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+
+/**
+ * @author Gail Badner
+ */
+@TestForIssue( jiraKey = "HHH-12729")
+public class HibernateFirstResultMaxResultsTest extends BaseNonConfigCoreFunctionalTestCase {
+
+	protected Class[] getAnnotatedClasses() {
+		return new Class[] { Employee.class };
+	}
+
+	@Test
+	public void testFirstResult() {
+		doInHibernate(
+				this::sessionFactory,
+				session -> {
+
+					Query query = session.createQuery( "from Employee" );
+
+					// not initialized yet
+					assertNull( query.getHibernateFirstResult() );
+
+					// the following is special case; when initialized to -1, getHibernateFirstResult returns 0
+					assertEquals( Integer.valueOf( 0 ), query.setHibernateFirstResult( -1 ).getHibernateFirstResult() );
+
+					assertEquals( Integer.valueOf( 0 ), query.setHibernateFirstResult( 0 ).getHibernateFirstResult() );
+					assertEquals( Integer.valueOf( 1 ), query.setHibernateFirstResult( 1 ).getHibernateFirstResult() );
+
+					assertEquals( Integer.valueOf( 10 ), query.setFirstResult( 10 ).getHibernateFirstResult() );
+				}
+		);
+	}
+
+	@Test
+	public void testMaxResults() {
+		doInHibernate(
+				this::sessionFactory,
+				session -> {
+
+					Query query = session.createQuery( "from Employee" );
+
+					// not initialized yet
+					assertNull( query.getHibernateMaxResults() );
+
+					// values <= 0 are considered uninitialized;
+					assertNull( query.setHibernateMaxResults( -1 ).getHibernateMaxResults() );
+					assertNull( query.setHibernateMaxResults( 0 ).getHibernateMaxResults() );
+
+					assertEquals( Integer.valueOf( 1 ), query.setHibernateMaxResults( 1 ).getHibernateMaxResults() );
+
+					assertEquals( Integer.valueOf( 0 ), query.setMaxResults( 0 ).getHibernateMaxResults() );
+
+					assertEquals( Integer.valueOf( 2 ), query.setMaxResults( 2 ).getHibernateMaxResults() );
+				}
+		);
+	}
+
+	@Test
+	public void testPagination() {
+		doInHibernate(
+				this::sessionFactory,
+				session -> {
+					for ( int i = 0; i < 5; i++ ) {
+						session.persist( new Employee( i ) );
+					}
+				}
+		);
+
+		final String query = "from Employee order by id";
+		checkResults( executeQuery( query, null, null ), 0, 4 );
```

Comment:
Could we add some tests about the new API rightfully interpreting the "maxresults=0" case as a need to return zero results?

Reply:
Don't we already have some somewhere else - I suppose they have been added when the new API was introduced? This test is about validating the compatibility, I don't think we should add tests about the new API here.

Reply:
Agreed it's orthogonal, yet this whole mess of needing to intepret zero as all is caused by that being the semantics before. Whish is clearly wrong, and I'd like us to make sure it never happens again.

(So I'm not sure if that's tested anywhere else as clearly we had it wrong)

+1 to manage it as a follow up JIRA if that's easier, didn't mean to "block" this one.

Reply:
In addition to https://github.com/scottmarlow/wildfly/tree/WFLY-10520_hibernatecompatabilitytransformer, in the jira description of https://issues.jboss.org/browse/WFLY-10520, we are trying to keep current, the list of transformations to make.

Reply:
There are a couple of tests that show that ```javax.persistence.Query#setMaxResults``` works as it should. Here are tests for ```setMaxResults( 0 )```:
* using CriteriaQuery -- [LimitExpressionTest](https://github.com/hibernate/hibernate-orm/blob/master/hibernate-core/src/test/java/org/hibernate/jpa/test/criteria/limitExpression/LimitExpressionTest.java#L36-L45)
* using Query -- [LimitExpressionTest](https://github.com/hibernate/hibernate-orm/blob/master/hibernate-core/src/test/java/org/hibernate/test/legacy/LimitExpressionTest.java#L33-L51)